cmpserver/plugin/plugin.go

```diff
@@ -32,11 +36,19 @@ func NewService(initConstants CMPServerInitConstants) *Service {
 	}
 }
 
-func runCommand(command Command, path string, env []string) (string, error) {
+func runCommand(ctx context.Context, command Command, path string, env []string) (string, error) {
 	if len(command.Command) == 0 {
 		return "", fmt.Errorf("Command is empty")
 	}
-	cmd := exec.Command(command.Command[0], append(command.Command[1:], command.Args...)...)
+	cmd := exec.CommandContext(ctx, command.Command[0], append(command.Command[1:], command.Args...)...)
+
+	// Make sure the command is killed immediately on timeout. https://stackoverflow.com/a/29552044/684776
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	go func() {
+		<-ctx.Done()
+		_ = syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
+	}()
```

Comment:
I think you want to use the process group ID instead of the process ID as argument to `syscall.Kill()` here, because you specify it as negative number, or specify just the command's PID as positive number.

From the manpage to `kill(2)`:

```
       The kill() system call can be used to send any signal to any process group or process.

       If pid is positive, then signal sig is sent to the process with the ID specified by pid.

       If pid equals 0, then sig is sent to every process in the process group of the calling process.

       If  pid  equals  -1,  then  sig is sent to every process for which the calling process has permission to send signals, except for
       process 1 (init), but see below.

       If pid is less than -1, then sig is sent to every process in the process group whose ID is -pid.
```

Reply:
Good catch. I'm actually not sure why this worked with the process ID. Will change and test with the group ID.

Reply:
Oh! The linked solution didn't work for me, that's why I went with this one. I just pasted the wrong link. https://stackoverflow.com/a/38133948/684776

Reply:
I had to do some reading, but I think I finally understand what this code is actually doing. From the [syscall docs](https://pkg.go.dev/syscall#SysProcAttr):

	// Setpgid sets the process group ID of the child to Pgid,
	// or, if Pgid == 0, to the new child's process ID.
	Setpgid bool

Since I'm not setting Pgid, the group ID is the same as the process ID.

Using the process ID saves a call to `syscall.Getpgid`, which returns an error that would have to be handled.

Reply:
Locally (in osx) the commands started by `sh -c` don't seem to run in the same group as `sh`. So I'm not sure what benefit killing the `sh` group would provide.

I'm testing again with just killing the `sh` process.

Reply:
With the go func commented out, the CMP server doesn't send a response immediately. This makes me think that the sub-process(es) aren't killed along with the root process.

I'll try again but kill the PID instead of the GID.

Reply:
Alright. I've developed a thousand-yard stare tinkering with this.

Running locally, I don't need this custom kill logic. The process is always killed immediately at timeout.

Running in Docker, the following combinations all fail to kill the process at timeout:

```go
	// no additional kill logic
```

```go
	// just setting the group id
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
```

```go
	// setting the group id and killing by process id
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	go func() {
		<-ctx.Done()
		_ = syscall.Kill(cmd.Process.Pid, syscall.SIGKILL)
	}()
```

This is the only combination that worked:

```go
	// kill the process by group id (which will always be the same as the process id)
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	go func() {
		<-ctx.Done()
		_ = syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
	}()
```

I've added a unit test and a comment about pid==gid. I have no idea if this issue will present in other environments. But I'm confident the extra code won't do any harm.